keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -265,31 +266,45 @@ class ApplicationService(
           potentialSummary
         }
       }
-      DEPLOYING -> {
-        val olderGitMetadata = repository.getGitMetadataByPromotionStatus(deliveryConfig, environmentName, artifact, CURRENT.name)
+
+      DEPLOYING, APPROVED -> {
+        val olderGitMetadata = pinnedArtifact?: repository.getArtifactVersionByPromotionStatus(deliveryConfig, environmentName, artifact, CURRENT.name)
         potentialSummary?.copy(
-          // comparing DEPLOYING (version in question, new code) vs. CURRENT (old code)
-          compareLink = generateDiffLink(baseScmUrl, artifactGitMetadata, olderGitMetadata)
+          // comparing DEPLOYING/APPROVED (version in question, new code) vs. CURRENT (old code)
+          compareLink = getUrl(currentArtifact, olderGitMetadata, artifact)
         )
       }
       PREVIOUS -> {
-        val newerGitMetadata = potentialSummary?.replacedBy?.let { getArtifactInstance(artifact, it)?.gitMetadata }
+        val newerGitMetadata = potentialSummary?.replacedBy?.let { getArtifactInstance(artifact, it) }
         potentialSummary?.copy(
           //comparing PREVIOUS (version in question, old code) vs. the version which replaced it (new code)
-          compareLink = generateDiffLink(baseScmUrl, newerGitMetadata, artifactGitMetadata)
+          //pinned artifact should not be consider here, as we know exactly which version replace the current one
+          compareLink = getUrl(currentArtifact, newerGitMetadata, artifact)
         )
       }
       CURRENT -> {
-        val olderGitMetadata = repository.getGitMetadataByPromotionStatus(deliveryConfig, environmentName, artifact, PREVIOUS.name)
+        val olderGitMetadata = pinnedArtifact?: repository.getArtifactVersionByPromotionStatus(deliveryConfig, environmentName, artifact, PREVIOUS.name)
         potentialSummary?.copy(
           // comparing CURRENT (version in question, new code) vs. PREVIOUS (old code)
-          compareLink = generateDiffLink(baseScmUrl, artifactGitMetadata, olderGitMetadata)
+          compareLink = getUrl(currentArtifact, olderGitMetadata, artifact)
         )
       }
       else -> potentialSummary
     }
   }
 
+  // Pinning is a special case when is coming to creating a compare link between versions.
+  // If there is a pinned version, which is not the same as the current version, we need
+  // to make sure we are creating the comparable link with reference to the pinned version.
+  private fun getPinnedArtifact(deliveryConfig: DeliveryConfig, environmentName: String, artifact: DeliveryArtifact, version: String): PublishedArtifact? {
+    val pinnedVersion = repository.getPinnedVersion(deliveryConfig, environmentName, artifact.reference)
+     return if (pinnedVersion != version)
```

Comment:
I'd check for null in this conditional so you don't have to do the `?` below. I think it's a bit more readable

Reply:
done

Reply:
I ended up not doing it as it got this method to be much longer (if `pinnedVersion` is null, I need to return null. Here its happening automatically, and if I were to add your suggestion, I needed to add a special case for it and it would have gotten less readable, in my opinion).